### PR TITLE
[GraphBolt] Optimize hetero sampling on CPU

### PR DESCRIPTION
**Pull request by RamonZhou**
## Description
The CPU part of #7223.

What has changed:
- Hetero sampling when fanouts.size() > 0

What hasn't changed:
- Temporal sampling (the effect is unsure).
- Hetero graph but fanouts.size() = 1 (this code change doesn't apply to this situation sufficiently. It needs to be implemented in another way).
- Homo sampling (shouldn't be changed).

The performance will be tested soon.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Review by mfbalin — COMMENTED** [review]
subgraph_type_per_edge output not needed for the new code path. So this tensor should be returned only when the old code path is used.

**Comment by RamonZhou** [comment]
> subgraph_type_per_edge output not needed for the new code path. So this tensor should be returned only when the old code path is used.

Sure when the new code path is used, `subgraph_type_per_edge` will be `torch::nullopt`. Except that, when hetero graph is sampled as a homo graph, it still uses the old code logic (generate `subgraph_type_per_edge` not `edge_offsets`) because I think it needs to be handled differently. Maybe we need to sort all the edges after sampling process since we do not know in advance that how many edges per each etype will be sampled.

**Comment by mfbalin** [comment]
> > subgraph_type_per_edge output not needed for the new code path. So this tensor should be returned only when the old code path is used.
> 
> Sure when the new code path is used, `subgraph_type_per_edge` will be `torch::nullopt`. Except that, when hetero graph is sampled as a homo graph, it still uses the old code logic (generate `subgraph_type_per_edge` not `edge_offsets`) because I think it needs to be handled differently. Maybe we need to sort all the edges after sampling process since we do not know in advance that how many edges per each etype will be sampled.

I think that is exactly what I did.

**Comment by mfbalin** [comment]
I will move the rest of the operations in the convert_to_sampled_subgraph into sample_neighbors soon. Such as subtracting the node offset from indices tensor and creating indptr tensor already ready to be sliced without the need for adjusting first element to be 0. FYI.

**Comment by mfbalin** [comment]
@RamonZhou #7312 includes the changes I mentioned yesterday.

File: graphbolt/src/fused_csc_sampling_graph.cc [hunk 1]
```diff
@@ -484,6 +492,238 @@ auto GetTemporalPickFn(
...
+
+                      const auto picked_number = pick_fn(
+                          offset, num_neighbors, picked_eids_data_ptr,
+                          seed_offset, subgraph_indptr_data_ptr,
+                          etype_id_to_num_picked_offset);
+
+                      // Step 5. Calculate other attributes and return the
+                      // subgraph.
+                      if (picked_number > 0) {
+                        AT_DISPATCH_INDEX_TYPES(
```
Comment (peizhou001) [inline graphbolt/src/fused_csc_sampling_graph.cc:666]: Is this dispatch used?

**Comment by peizhou001** [comment]
LGTM for this PR. My concern revolves around temporal sampling:

1. Has temporal sampling been adapted for this corresponding change?
2. While most of the code is shared between normal sampling and temporal sampling, it's structured in a flattened manner within a single function, making it less reusable.

**Comment by mfbalin** [comment]
> @RamonZhou #7312 includes the changes I mentioned yesterday.

It is now merged so you might need to modify your code slightly after merging changes from master to be able to pass the CI. The indptr change is as follows:

if we were outputting 0 2 4 7 9 previously with 0 2 4 being indptr of first etype and 4 7 9 being the indptr for the second etype, now we duplicate the middle 4 to get 0 2 4 4 7 9, then we subtract 4 for the 2nd etype so that its indptr starts from 0. Thus, final output becomes:

0 2 4 0 3 5

For the indices, we make sure to subtract the source node type offset from the output indices tensor. You can reverse engineer these changes by looking at the `_convert_to_sampled_subgraph` changes in #7312.

**Comment by RamonZhou** [comment]
> > @RamonZhou #7312 includes the changes I mentioned yesterday.
> 
> It is now merged so you might need to modify your code slightly after merging changes from master to be able to pass the CI. The indptr change is as follows:
> 
> if we were outputting 0 2 4 7 9 previously with 0 2 4 being indptr of first etype and 4 7 9 being the indptr for the second etype, now we duplicate the middle 4 to get 0 2 4 4 7 9, then we subtract 4 for the 2nd etype so that its indptr starts from 0. Thus, final output becomes:
> 
> 0 2 4 0 3 5
> 
> For the indices, we make sure to subtract the source node type offset from the output indices tensor. You can reverse engineer these changes by looking at the `_convert_to_sampled_subgraph` changes in #7312.

I see I will work on it. But what should the `original_column_ids` be in this case? Should it still be the same length as indptr?

**Comment by mfbalin** [comment]
> > > @RamonZhou #7312 includes the changes I mentioned yesterday.
> > 
> > 
> > It is now merged so you might need to modify your code slightly after merging changes from master to be able to pass the CI. The indptr change is as follows:
> > if we were outputting 0 2 4 7 9 previously with 0 2 4 being indptr of first etype and 4 7 9 being the indptr for the second etype, now we duplicate the middle 4 to get 0 2 4 4 7 9, then we subtract 4 for the 2nd etype so that its indptr starts from 0. Thus, final output becomes:
> > 0 2 4 0 3 5
> > For the indices, we make sure to subtract the source node type offset from the output indices tensor. You can reverse engineer these changes by looking at the `_convert_to_sampled_subgraph` changes in #7312.
> 
> I see I will work on it. But what should the `original_column_ids` be in this case? Should it still be the same length as indptr?

It does not need to change at all. In the GPU implementation, I simply return the seeds that were passed in as input.

**Comment by RamonZhou** [comment]
> > > > @RamonZhou #7312 includes the changes I mentioned yesterday.
> > > 
> > > 
> > > It is now merged so you might need to modify your code slightly after merging changes from master to be able to pass the CI. The indptr change is as follows:
> > > if we were outputting 0 2 4 7 9 previously with 0 2 4 being indptr of first etype and 4 7 9 being the indptr for the second etype, now we duplicate the middle 4 to get 0 2 4 4 7 9, then we subtract 4 for the 2nd etype so that its indptr starts from 0. Thus, final output becomes:
> > > 0 2 4 0 3 5
> > > For the indices, we make sure to subtract the source node type offset from the output indices tensor. You can reverse engineer these changes by looking at the `_convert_to_sampled_subgraph` changes in #7312.
> > 
> > 
> > I see I will work on it. But what should the `original_column_ids` be in this case? Should it still be the same length as indptr?
> 
> It does not need to change at all. In the GPU implementation, I simply return the seeds that were passed in as input.

The changes will actually affect the distributed part of DGL where the code uses `original_column_ids` to restore the graph information.

**Comment by RamonZhou** [comment]
> > > > > > @RamonZhou #7312 includes the changes I mentioned yesterday.
> > > > > 
> > > > > 
> > > > > It is now merged so you might need to modify your code slightly after merging changes from master to be able to pass the CI. The indptr change is as follows:
> > > > > if we were outputting 0 2 4 7 9 previously with 0 2 4 being indptr of first etype and 4 7 9 being the indptr for the second etype, now we duplicate the middle 4 to get 0 2 4 4 7 9, then we subtract 4 for the 2nd etype so that its indptr starts from 0. Thus, final output becomes:
> > > > > 0 2 4 0 3 5
> > > > > For the indices, we make sure to subtract the source node type offset from the output indices tensor. You can reverse engineer these changes by looking at the `_convert_to_sampled_subgraph` changes in #7312.
> > > > 
> > > > 
> > > > I see I will work on it. But what should the `original_column_ids` be in this case? Should it still be the same length as indptr?
> > > 
> > > 
> > > It does not need to change at all. In the GPU implementation, I simply return the seeds that were passed in as input.
> > 
> > 
> > The changes will actually affect the distributed part of DGL where the code uses `original_column_ids` to restore the graph information.
> 
> I believe distDGL does not pass a seed_offsets argument, so the new code path is not going to run. Otherwise, I think it is good to do the same things as the CUDA implementation.

I see. I actually changed the distDGL code to make it pass seed_offsets to apply the new changes to it. I can change it back so distDGL can remain unchanged and maybe we can do it in the future.

**Comment by RamonZhou** [comment]
@dgl-bot

File: graphbolt/include/graphbolt/fused_csc_sampling_graph.h [hunk 2]
```diff
@@ -498,13 +505,14 @@ class FusedCSCSamplingGraph : public torch::CustomClassHolder {
...
  * node.
  * @param num_neighbors The number of neighbors of this node.
- *
- * @return The pick number of the given node.
+ * @param num_picked_ptr The pointer of the tensor which stores the pick
+ * numbers.
  */
-int64_t NumPick(
+template <typename PickedNumType>
+void NumPick(
```
Comment (mfbalin) [inline graphbolt/include/graphbolt/fused_csc_sampling_graph.h:512]: Why do we need to change the return type here?

File: graphbolt/include/graphbolt/fused_csc_sampling_graph.h [hunk 3]
```diff
@@ -513,11 +521,13 @@ int64_t TemporalNumPick(
     const torch::optional<torch::Tensor>& edge_timestamp, int64_t seed_offset,
     int64_t offset, int64_t num_neighbors);
 
-int64_t NumPickByEtype(
-    const std::vector<int64_t>& fanouts, bool replace,
+template <typename PickedNumType>
+void NumPickByEtype(
```
Comment (mfbalin) [inline graphbolt/include/graphbolt/fused_csc_sampling_graph.h:525]: Ditto

File: graphbolt/include/graphbolt/fused_csc_sampling_graph.h [hunk 4]
```diff
@@ -610,16 +620,24 @@ int64_t TemporalPick(
...
+ * of the sampled subgraph.
+ * @param etype_id_to_num_picked_offset A vector storing the mappings from each
+ * etype_id to the offset of its pick numbers in the tensor.
  */
 template <SamplerType S, typename PickedType>
 int64_t PickByEtype(
-    int64_t offset, int64_t num_neighbors, const std::vector<int64_t>& fanouts,
-    bool replace, const torch::TensorOptions& options,
-    const torch::Tensor& type_per_edge,
+    bool with_seed_offsets, int64_t offset, int64_t num_neighbors,
```
Comment (mfbalin) [inline graphbolt/include/graphbolt/fused_csc_sampling_graph.h:634]: Can we move with_seed_offsets so that it is together with seed_offset?

File: tests/python/pytorch/graphbolt/impl/test_fused_csc_sampling_graph.py [hunk 5]
```diff
@@ -2219,10 +2219,13 @@ def test_sample_neighbors_hetero_pick_number(
         type_per_edge=type_per_edge,
         node_type_to_id=ntypes,
         edge_type_to_id=etypes,
-    )
+    ).to(F.ctx())
 
     # Generate subgraph via sample neighbors.
-    nodes = torch.LongTensor([0, 1])
+    nodes = {
```
Comment (mfbalin) [inline tests/python/pytorch/graphbolt/impl/test_fused_csc_sampling_graph.py:2225]: Is this change needed so that seed_offsets can be generated before sample neighbors is called?

Comment (mfbalin) [reply tests/python/pytorch/graphbolt/impl/test_fused_csc_sampling_graph.py:2225]: I wonder why the GPU tests passed without this change.

File: graphbolt/include/graphbolt/fused_csc_sampling_graph.h (hunk 4 above)
Comment (mfbalin) [reply graphbolt/include/graphbolt/fused_csc_sampling_graph.h:634]: Actually, it is better to use std::optional<int64_t> seed_offset so that we don't need the extra boolean.

File: graphbolt/src/fused_csc_sampling_graph.cc [hunk 6]
```diff
@@ -355,17 +356,23 @@ c10::intrusive_ptr<FusedSampledSubgraph> FusedCSCSamplingGraph::InSubgraph(
 auto GetNumPickFn(
     const std::vector<int64_t>& fanouts, bool replace,
     const torch::optional<torch::Tensor>& type_per_edge,
-    const torch::optional<torch::Tensor>& probs_or_mask) {
+    const torch::optional<torch::Tensor>& probs_or_mask,
+    bool with_seed_offsets) {
```
Comment (mfbalin) [inline graphbolt/src/fused_csc_sampling_graph.cc:360]: if we use a torch::optional<int64_t> for seed_offset, this boolean is not needed.

File: graphbolt/src/fused_csc_sampling_graph.cc [hunk 7]
```diff
@@ -423,21 +430,25 @@ auto GetPickFn(
...
     const torch::TensorOptions& options,
     const torch::optional<torch::Tensor>& type_per_edge,
-    const torch::optional<torch::Tensor>& probs_or_mask, SamplerArgs<S> args) {
-  return [&fanouts, replace, &options, &type_per_edge, &probs_or_mask, args](
-             int64_t seed_offset, int64_t offset, int64_t num_neighbors,
-             auto picked_data_ptr) {
+    const torch::optional<torch::Tensor>& probs_or_mask, bool with_seed_offsets,
+    SamplerArgs<S> args) {
+  return [&fanouts, replace, &options, &type_per_edge, &probs_or_mask, args,
+          with_seed_offsets](
```
Comment (mfbalin) [inline graphbolt/src/fused_csc_sampling_graph.cc:436]: ditto

File: graphbolt/src/fused_csc_sampling_graph.cc [hunk 8]
```diff
@@ -484,6 +495,309 @@ auto GetTemporalPickFn(
 template <typename NumPickFn, typename PickFn>
 c10::intrusive_ptr<FusedSampledSubgraph>
 FusedCSCSamplingGraph::SampleNeighborsImpl(
+    const torch::Tensor& seeds,
+    torch::optional<std::vector<int64_t>> seed_offsets,
```
Comment (mfbalin) [inline graphbolt/src/fused_csc_sampling_graph.cc:499]:
```suggestion
    torch::optional<std::vector<int64_t>>& seed_offsets,
```

File: graphbolt/src/fused_csc_sampling_graph.cc [hunk 9]
```diff
@@ -484,6 +495,309 @@ auto GetTemporalPickFn(
...
+  }
+
+  return c10::make_intrusive<FusedSampledSubgraph>(
+      subgraph_indptr, subgraph_indices, seeds, torch::nullopt,
+      subgraph_reverse_edge_ids, subgraph_type_per_edge, edge_offsets);
+}
+
+template <typename NumPickFn, typename PickFn>
+c10::intrusive_ptr<FusedSampledSubgraph>
+FusedCSCSamplingGraph::TemporalSampleNeighborsImpl(
```
Comment (mfbalin) [inline graphbolt/src/fused_csc_sampling_graph.cc:800]: Why did we need to copy the whole SampleNeighborsImpl code?

Comment (mfbalin) [reply graphbolt/src/fused_csc_sampling_graph.cc:800]: If there is a deep reason, we can merge this PR with duplicated code and open an issue to refactor it so that the two code paths are same again.

File: graphbolt/src/fused_csc_sampling_graph.cc [hunk 10]
```diff
@@ -663,18 +977,22 @@ c10::intrusive_ptr<FusedSampledSubgraph> FusedCSCSamplingGraph::SampleNeighbors(
     }
   }
 
+  bool with_seed_offsets = seed_offsets.has_value();
```
Comment (mfbalin) [inline graphbolt/src/fused_csc_sampling_graph.cc:980]: We don't need this boolean if we use torch::optional for the seed offset variables.

**Comment by mfbalin** [comment]
@RamonZhou Do we have to use an extra boolean called with_seed_offsets, could it also work if we made seed_offset variable into an torch::optional<int64_t> itself?

File: graphbolt/src/fused_csc_sampling_graph.cc [hunk 11]
```diff
@@ -484,6 +495,309 @@ auto GetTemporalPickFn(
...
+                        const auto seed_type_id =
+                            (fanouts.size() > 1) ? std::upper_bound(
+                                                       seed_offsets->begin(),
+                                                       seed_offsets->end(), i) -
+                                                       seed_offsets->begin() - 1
+                                                 : 0;
+                        // `seed_offset` indicates the index of the current
+                        // seed within the group of seeds which have the same
+                        // node type.
+                        const auto seed_offset =
```
Comment (mfbalin) [inline graphbolt/src/fused_csc_sampling_graph.cc:597]: A better name would be more appropriate, right now it is same as the seed_offsets tensor.

File: graphbolt/src/fused_csc_sampling_graph.cc [hunk 12]
```diff
@@ -484,6 +495,309 @@ auto GetTemporalPickFn(
...
+              auto picked_eids_data_ptr = picked_eids.data_ptr<indptr_t>();
+              torch::parallel_for(
+                  0, num_seeds, grain_size, [&](int64_t begin, int64_t end) {
+                    for (int64_t i = begin; i < end; ++i) {
+                      const auto nid = seeds_data_ptr[i];
+                      const auto offset = indptr_data[nid];
+                      const auto num_neighbors = indptr_data[nid + 1] - offset;
+                      auto picked_number = 0;
+                      const auto seed_type_id =
+                          (with_seed_offsets && fanouts.size() > 1)
```
Comment (mfbalin) [inline graphbolt/src/fused_csc_sampling_graph.cc:686]: let's refactor `with_seed_offsets && fanouts.size() > 1` into a variable as it is used commonly.

File: tests/python/pytorch/graphbolt/impl/test_fused_csc_sampling_graph.py (hunk 5 above)
Comment (RamonZhou) [reply tests/python/pytorch/graphbolt/impl/test_fused_csc_sampling_graph.py:2225]: Yes it's needed to generate seed_offsets. I guess this test wasn't run on GPU entirely because the graph wasn't carried to GPU so it didn't go into the CUDA branch.

**Comment by mfbalin** [comment]
LGTM overall. I don't like that we are replicating most of the code, however it could be addressed in a seperate PR.

File: tests/python/pytorch/graphbolt/impl/test_fused_csc_sampling_graph.py (hunk 5 above)
Comment (mfbalin) [reply tests/python/pytorch/graphbolt/impl/test_fused_csc_sampling_graph.py:2225]: Good catch.

File: graphbolt/src/fused_csc_sampling_graph.cc (hunk 9 above)
Comment (RamonZhou) [reply graphbolt/src/fused_csc_sampling_graph.cc:800]: I tried to use template function to merge them but the combination of `if constexpr` and lambda function somehow triggers the bug of MSVC so it couldn't build on windows. I didn't find a way to bypass it so I copied the code. We can try to find a nicer way to merge in another PR.

File: graphbolt/src/fused_csc_sampling_graph.cc (hunk 1 above)
Comment (RamonZhou) [reply graphbolt/src/fused_csc_sampling_graph.cc:719]: Yes it's used in line 723

File: graphbolt/src/fused_csc_sampling_graph.cc (hunk 9 above)
Comment (mfbalin) [reply graphbolt/src/fused_csc_sampling_graph.cc:800]: I see. That is unfortunate. Maybe we could try a later MSVC version? We should note this down.

Comment (mfbalin) [reply graphbolt/src/fused_csc_sampling_graph.cc:800]: I know you don't have a lot of time left but if you can open a PR after this is merged that has the refactored version, we can keep it as reference for later. If you can't do that, it should be okay.

Comment (RamonZhou) [reply graphbolt/src/fused_csc_sampling_graph.cc:800]: Yes that's a good idea

File: graphbolt/src/fused_csc_sampling_graph.cc (hunk 11 above)
Comment (RamonZhou) [reply graphbolt/src/fused_csc_sampling_graph.cc]: Renamed to `seed_index`

File: graphbolt/src/fused_csc_sampling_graph.cc [hunk 13]
```diff
@@ -484,6 +495,299 @@ auto GetTemporalPickFn(
...
+  // (hetero graph but sampled as a homo graph), set num_etypes as 1.
+  // In temporal sampling, this will not be used for now since the logic hasn't
+  // been adopted for temporal sampling.
+  const int64_t num_etypes =
+      (edge_type_to_id_.has_value() && hetero_with_seed_offsets)
+          ? edge_type_to_id_->size()
+          : 1;
+  std::vector<int64_t> etype_id_to_src_ntype_id(num_etypes);
+  std::vector<int64_t> etype_id_to_dst_ntype_id(num_etypes);
+  torch::Tensor subgraph_indptr_substract;
```
Comment (mfbalin) [inline graphbolt/src/fused_csc_sampling_graph.cc:527]: Let's make this torch::optional<torch::Tensor>, because it may not have a value.

File: graphbolt/src/fused_csc_sampling_graph.cc [hunk 14]
```diff
@@ -484,6 +495,299 @@ auto GetTemporalPickFn(
...
+      auto dst_ntype_id = node_type_to_id_->at(dst_type);
+      etype_id_to_src_ntype_id[id] = node_type_to_id_->at(src_type);
+      etype_id_to_dst_ntype_id[id] = dst_ntype_id;
+      etype_id_to_num_picked_offset[id + 1] =
+          seed_offsets->at(dst_ntype_id + 1) - seed_offsets->at(dst_ntype_id) +
+          1;
+    }
+    for (int64_t i = 1; i <= num_etypes; ++i) {
+      etype_id_to_num_picked_offset[i] += etype_id_to_num_picked_offset[i - 1];
+    }
```
Comment (mfbalin) [inline graphbolt/src/fused_csc_sampling_graph.cc:549]:
```suggestion
    std::partial_sum(etype_id_to_num_picked_offset.begin(), etype_id_to_num_picked_offset.end(), etype_id_to_num_picked_offset.begin());
```

File: graphbolt/src/fused_csc_sampling_graph.cc [hunk 15]
```diff
@@ -484,6 +495,299 @@ auto GetTemporalPickFn(
...
+  }
+  // `num_rows` indicates the length of `num_picked_neighbors_per_node`, which
+  // is used for storing pick numbers. In non-temporal hetero sampling, it
+  // equals to sum_{etype} #seeds with ntype=dst_type(etype). In homo sampling,
+  // it equals to `num_seeds`.
+  const int64_t num_rows = (fanouts.size() == 1 || !with_seed_offsets)
+                               ? num_seeds + 1
+                               : etype_id_to_num_picked_offset[num_etypes];
+  torch::Tensor num_picked_neighbors_per_node =
+      torch::zeros({num_rows}, indptr_options);
```
Comment (mfbalin) [inline graphbolt/src/fused_csc_sampling_graph.cc:562]:
```suggestion
      torch::empty({num_rows}, indptr_options);
```

File: graphbolt/src/fused_csc_sampling_graph.cc [hunk 16]
```diff
@@ -484,6 +495,299 @@ auto GetTemporalPickFn(
...
+    }
+  } else {
+    etype_id_to_dst_ntype_id[0] = 0;
+    etype_id_to_num_picked_offset[1] = num_seeds + 1;
+  }
+  // `num_rows` indicates the length of `num_picked_neighbors_per_node`, which
+  // is used for storing pick numbers. In non-temporal hetero sampling, it
+  // equals to sum_{etype} #seeds with ntype=dst_type(etype). In homo sampling,
+  // it equals to `num_seeds`.
+  const int64_t num_rows = (fanouts.size() == 1 || !with_seed_offsets)
```
Comment (mfbalin) [inline graphbolt/src/fused_csc_sampling_graph.cc:558]: The condition here does not seem to be needed as you already initialize `etype_id_to_num_picked_offset[1]` to `num_seeds + 1`.

Comment (mfbalin) [reply graphbolt/src/fused_csc_sampling_graph.cc:558]: You should be able to simply use `num_rows = etype_id_to_num_picked_offset[num_etypes]`.

File: graphbolt/src/fused_csc_sampling_graph.cc (hunk 15 above)
Comment (mfbalin) [reply graphbolt/src/fused_csc_sampling_graph.cc:562]: Is there any entry that is not initialized below? If so, we can use empty.

File: graphbolt/src/fused_csc_sampling_graph.cc [hunk 17]
```diff
@@ -484,6 +495,299 @@ auto GetTemporalPickFn(
...
+              // Step 2. Calculate prefix sum to get total length and offsets of
+              // each node. It's also the indptr of the generated subgraph.
+              subgraph_indptr = num_picked_neighbors_per_node.cumsum(
+                  0, indptr_.scalar_type());
+              auto subgraph_indptr_data_ptr =
+                  subgraph_indptr.data_ptr<indptr_t>();
+
+              // When doing non-temporal hetero sampling, we generate an
+              // edge_offsets tensor.
+              if (type_per_edge_.has_value() && hetero_with_seed_offsets) {
```
Comment (mfbalin) [inline graphbolt/src/fused_csc_sampling_graph.cc:645]: Do we need type_per_edge_ check here? If hetero_with_seed_offsets is true, then it should already be a valid tensor.

File: graphbolt/src/fused_csc_sampling_graph.cc [hunk 18]
```diff
@@ -484,6 +495,299 @@ auto GetTemporalPickFn(
...
+                                for (auto j = picked_begin; j < picked_end;
+                                     ++j) {
+                                  subgraph_indices_data_ptr[j] =
+                                      indices_data_ptr[picked_eids_data_ptr[j]];
+                                  if (hetero_with_seed_offsets &&
+                                      node_type_offset_.has_value())
+                                    // Substract the node type offset from
+                                    // subgraph indices.
+                                    AT_DISPATCH_INDEX_TYPES(
+                                        node_type_offset_.value().scalar_type(),
```
Comment (mfbalin) [inline graphbolt/src/fused_csc_sampling_graph.cc:735]: do we need a dispatch here? Doesn't the node_type_offset match the indices dtype already?

File: graphbolt/src/fused_csc_sampling_graph.cc (hunk 15 above)
Comment (RamonZhou) [reply graphbolt/src/fused_csc_sampling_graph.cc:562]: Yes it's because of the duplicates. Some positions might not be initialized.

File: graphbolt/src/fused_csc_sampling_graph.cc (hunk 18 above)
Comment (RamonZhou) [reply graphbolt/src/fused_csc_sampling_graph.cc:735]: Is this always the case? I wasn't sure about that so I added a dispatch.

File: graphbolt/src/fused_csc_sampling_graph.cc (hunk 15 above)
Comment (mfbalin) [reply graphbolt/src/fused_csc_sampling_graph.cc:562]: It is possible to manually set the duplicate positions that will be uninitialized to 0. Then we can use empty safely. We can either place a todo note or if you think it is easy enough, you can do this optimization.

File: graphbolt/src/fused_csc_sampling_graph.cc (hunk 18 above)
Comment (mfbalin) [reply graphbolt/src/fused_csc_sampling_graph.cc:741]: You can try removing it and see if all the tests pass locally rather quickly. I run graphbolt tests only for quick testing before pushing code.

**Comment by RamonZhou** [comment]
@dgl-bot

**Comment by RamonZhou** [comment]
Because the CI on windows was down, we need to move this to a new PR.

**Comment by mfbalin** [comment]
Can we resolve the conversations here so that we don't forget anything? If you resolved a comment, feel free to click on resolve conversation.
